File: GH_Inj_Lib_DotNet/Inj_VB_Example/SymbolsDownloader.vb
Imports System.ComponentModel
Imports GH_Inj_Lib_DotNet

Public Class SymbolsDownloader

    Public Property EngineDLL As String = String.Empty
    Public Property Is64 As Boolean = False
    Private DI As New Core.DynamicInvoke

    Private IsFinalized As Boolean = False

    Private Sub SymbolsDownloader_Closing(sender As Object, e As CancelEventArgs) Handles Me.Closing
        If IsFinalized = False Then
            If MessageBox.Show("Are you sure you want to cancel the download?", "Symbols Downloader", MessageBoxButtons.YesNo) = DialogResult.Yes Then
                DI.Invoke("InterruptDownload", EngineDLL, GetType(Boolean), Nothing)
                System.Threading.Thread.Sleep(500)
                Me.DialogResult = DialogResult.Cancel
            Else
                e.Cancel = True
            End If
        End If
    End Sub

    Private Sub SymbolsDownloader_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        Label1.Text = "Downloading Symbols... "
    End Sub

    Private Sub SymbolsDownloader_Shown(sender As Object, e As EventArgs) Handles Me.Shown
        DI.Invoke("StartDownload", EngineDLL, GetType(Boolean), Nothing)
        System.Threading.Thread.Sleep(500)
        Timer1.Enabled = True
    End Sub

    Private Sub Timer1_Tick(sender As Object, e As EventArgs) Handles Timer1.Tick
        Dim GetSymbolState As IntPtr = DI.Invoke("GetSymbolState", EngineDLL, GetType(IntPtr), Nothing)
        Dim GetImportState As IntPtr = DI.Invoke("GetImportState", EngineDLL, GetType(IntPtr), Nothing)

        Dim GetDownloadProgress As Integer = Core.GHWrapper.GetDownloadProgressEx(Is64)
        ProgressBar1.Value = GetDownloadProgress

        If GetSymbolState.ToInt32 = 0 Then
            'SymbolState Loaded
        End If

        If GetImportState.ToInt32 = 0 Then
            'ImportState Loaded
            IsFinalized = True
            Timer1.Enabled = False
            Me.DialogResult = DialogResult.OK
            Me.Close()
        End If

    End Sub
End Class

File: GH_Inj_Lib_DotNet/Inj_CSharp_Example/SymbolsDownloader.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Inj_CSharp_Example
{
    public partial class SymbolsDownloader : Form
    {

        public string EngineDLL { get; set; } = string.Empty;
        public bool Is64 { get; set; } = false;
        private GH_Inj_Lib_DotNet.Core.DynamicInvoke DI = new GH_Inj_Lib_DotNet.Core.DynamicInvoke();

        private bool IsFinalized = false;

        private void SymbolsDownloader_Closing(object sender, CancelEventArgs e)
        {
            if (IsFinalized == false)
            {
                if (MessageBox.Show("Are you sure you want to cancel the download?", "Symbols Downloader", MessageBoxButtons.YesNo) == DialogResult.Yes)
                {
                    DI.Invoke("InterruptDownload", EngineDLL, typeof(bool), null/* TODO Change to default(_) if this is not a reference type */);
                    System.Threading.Thread.Sleep(500);
                    this.DialogResult = DialogResult.Cancel;
                }
                else
                    e.Cancel = true;
            }
        }


        public SymbolsDownloader()
        {
            InitializeComponent();
        }

        private void SymbolsDownloader_Load(object sender, EventArgs e)
        {
           Label1.Text = "Downloading Symbols... ";
        }

        private void SymbolsDownloader_Shown(object sender, EventArgs e)
        {
            DI.Invoke("StartDownload", EngineDLL, typeof(bool), null/* TODO Change to default(_) if this is not a reference type */);
            System.Threading.Thread.Sleep(500);
            Timer1.Enabled = true;
        }

        private void Timer1_Tick(object sender, EventArgs e)
        {
            IntPtr GetSymbolState = (IntPtr)DI.Invoke("GetSymbolState", EngineDLL, typeof(IntPtr), null);
            IntPtr GetImportState = (IntPtr)DI.Invoke("GetImportState", EngineDLL, typeof(IntPtr), null);

            int GetDownloadProgress = GH_Inj_Lib_DotNet.Core.GHWrapper.GetDownloadProgressEx(Is64);
            ProgressBar1.Value = GetDownloadProgress;

            if (GetSymbolState.ToInt32() == 0)
            {
            }

            if (GetImportState.ToInt32() == 0)
            {
                // ImportState Loaded
                IsFinalized = true;
                Timer1.Enabled = false;
                this.DialogResult = DialogResult.OK;
                this.Close();
            }
        }

    }
}
